src/Evolution/Systems/GrMhd/ValenciaDivClean/Actions/NumericInitialData.hpp

```diff
@@ -23,103 +24,269 @@
+
+  template <typename... AllTags>
+  void set_initial_data(
+      const gsl::not_null<Scalar<DataVector>*> rest_mass_density,
+      const gsl::not_null<Scalar<DataVector>*> electron_fraction,
+      const gsl::not_null<Scalar<DataVector>*> specific_internal_energy,
+      const gsl::not_null<tnsr::I<DataVector, 3>*> spatial_velocity,
+      const gsl::not_null<tnsr::I<DataVector, 3>*> magnetic_field,
+      const gsl::not_null<Scalar<DataVector>*> div_cleaning_field,
+      const gsl::not_null<Scalar<DataVector>*> lorentz_factor,
+      const gsl::not_null<Scalar<DataVector>*> pressure,
+      const gsl::not_null<Scalar<DataVector>*> specific_enthalpy,
+      const gsl::not_null<tuples::TaggedTuple<AllTags...>*> numeric_data,
+      const tnsr::II<DataVector, 3>& inv_spatial_metric,
+      const EquationsOfState::EquationOfState<true, 1>& equation_of_state)
+      const {
+    // Rest mass density from dataset
+    *rest_mass_density =
+        std::move(get<hydro::Tags::RestMassDensity<DataVector>>(*numeric_data));
+    const size_t num_points = get(*rest_mass_density).size();
+    // Electron fraction from dataset or constant value
+    const std::variant<double, std::string>& electron_fraction_selection =
+        get<VarName<hydro::Tags::ElectronFraction<DataVector>,
+                    std::bool_constant<false>>>(selected_variables_);
+    if (std::holds_alternative<std::string>(electron_fraction_selection)) {
+      *electron_fraction = std::move(
+          get<hydro::Tags::ElectronFraction<DataVector>>(*numeric_data));
+    } else {
+      const double constant_electron_fraction =
+          std::get<double>(electron_fraction_selection);
+      destructive_resize_components(electron_fraction, num_points);
+      get(*electron_fraction) = constant_electron_fraction;
+    }
+    // Velocity and Lorentz factor from u_i dataset
+    // W = 1 + W^2 v_i v^i
+    // where W v_i = u_i, so we first raise the index on W v_i with the
+    // spatial metric and then compute its magnitude. We use
+    // `spatial_velocity` as intermediate memory buffer for W v^i.
+    const auto& u_i = get<hydro::Tags::LowerSpatialFourVelocity<DataVector, 3>>(
+        *numeric_data);
+    raise_or_lower_index(spatial_velocity, u_i, inv_spatial_metric);
+    dot_product(lorentz_factor, u_i, *spatial_velocity);
+    get(*lorentz_factor) += 1.;
+    for (size_t d = 0; d < 3; ++d) {
+      spatial_velocity->get(d) /= get(*lorentz_factor);
+    }
+    // Specific internal energy, pressure, and enthalpy from EOS
+    destructive_resize_components(specific_internal_energy, num_points);
+    destructive_resize_components(pressure, num_points);
+    destructive_resize_components(specific_enthalpy, num_points);
+    for (size_t i = 0; i < num_points; ++i) {
+      double& local_rest_mass_density = get(*rest_mass_density)[i];
+      // Apply the EOS only where the density is above the cutoff, because
+      // the EOS may be ill-defined at zero density. Also, the specific
```

Comment:
It's not just the EOS, the fluid model itself breaks down in the zero-density limit. From stat mech, the fluid model assumes a large (Avogadro's number) of particles, which is not true in vacuum. 

Would it be possible to apply an actual atmosphere set in the input file? I think we currently just do that after the initial data is set, so that's fine, but I don't think you should call the treatment here "atmosphere" because it's fundamentally different from the atmosphere treatment.

Reply:
Yep, and I don't call this "atmosphere" :) The help string for the `DensityCutoff` is this:

> Where the density is below this cutoff the fluid variables are set to vacuum (zero density, pressure, energy and velocity, unit Lorentz factor and enthalpy). During the evolution, atmosphere treatment will typically kick in and fix the value of the fluid variables in these regions. Therefore, it makes sense to set this density cutoff to the same value as the atmosphere density cutoff.

Sound ok? I reworded the EOS comment a bit and added some docs.

Reply:
Yep, looks good!